File: api/routes/sims.py
from __future__ import annotations

import asyncio
import json
import os
from datetime import datetime, timezone
from typing import Annotated, Any

from fastapi import APIRouter, BackgroundTasks, File, Form, HTTPException, Path, Query, UploadFile
from fastapi.responses import FileResponse, StreamingResponse
from pydantic import BaseModel, Field

from models import RunMetadata
from services import engine_adapter
from services.runner import PROGRESS, run_simulation_job
from services.sim_artifacts import (
    analysis_xlsx_path,
    chart_path,
    godprogs_records,
    list_chart_filenames,
    player_all_runs,
    player_summaries,
    raw_outputs_csv_path,
)
from services.storage import delete_run, get_run, list_runs, outputs_root, repo_root

router = APIRouter()

BuildId = Annotated[
    str,
    Path(pattern=r"^\d{14}$", description="CalVer build id (YYYYMMDDHHmmss)"),
]


class SimCreateBody(BaseModel):
    teams: list[str] = Field(default_factory=list)
    seed: int = 69
    runs: int = 500
    n_workers: int | None = None


def _new_build_id() -> str:
    return datetime.now(timezone.utc).strftime("%Y%m%d%H%M%S")


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


@router.get("")
async def list_sims() -> list[RunMetadata]:
    """List runs from `outputs/` (newest first)."""
    return list_runs()


@router.post("")
async def create_sim(
    background_tasks: BackgroundTasks,
    export: UploadFile = File(..., description="League export JSON"),
    config: str = Form(..., description="JSON body: teams, seed, runs, n_workers"),
    teaminfo: UploadFile | None = File(default=None, description="Optional teaminfo.json"),
) -> dict[str, str]:
    """Upload export (+ optional teaminfo) and start a simulation."""
    body = SimCreateBody.model_validate_json(config)
    cpu = os.cpu_count() or 1
    n_workers = body.n_workers if body.n_workers is not None else max(cpu - 1, 1)
    if n_workers < 1:
        raise HTTPException(status_code=422, detail="n_workers must be >= 1")

    build = _new_build_id()
    out = outputs_root() / build
    out.mkdir(parents=True, exist_ok=True)

    export_bytes = await export.read()
    if not export_bytes:
        raise HTTPException(status_code=422, detail="export file is empty")

    export_path = out / "export.json"
    export_path.write_bytes(export_bytes)

    if teaminfo is not None:
        teaminfo_bytes = await teaminfo.read()
        if not teaminfo_bytes:
            raise HTTPException(status_code=422, detail="teaminfo file is empty")
    else:
        default_teaminfo = repo_root() / "data" / "teaminfo.json"
        if not default_teaminfo.is_file():
            raise HTTPException(
                status_code=500,
                detail="No teaminfo upload and data/teaminfo.json is missing",
            )
        teaminfo_bytes = default_teaminfo.read_bytes()

    teaminfo_path = out / "teaminfo.json"
    teaminfo_path.write_bytes(teaminfo_bytes)

    meta = {
        "build": build,
        "script_version": engine_adapter.script_version(),
        "teams": body.teams,
        "seed": body.seed,
        "runs": body.runs,
        "n_workers": n_workers,
        "export_file": f"outputs/{build}/export.json",
        "teaminfo_file": f"outputs/{build}/teaminfo.json",
        "status": "running",
        "started_at": _utc_now_iso(),
        "completed_at": None,
        "player_count": None,
        "config_snapshot": engine_adapter.config_snapshot(),
        "error": None,
    }
    (out / "metadata.json").write_text(json.dumps(meta, indent=2), encoding="utf-8")

    background_tasks.add_task(
        run_simulation_job,
        build,
        export_path,
        teaminfo_path,
        body.teams,
        body.seed,
        body.runs,
        n_workers,
    )
    return {"build": build}


@router.get("/{build}/progress")
async def sim_progress_sse(build: BuildId) -> StreamingResponse:
    """SSE stream: JSON lines `{ phase, pct, message, done }`."""
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")

    async def event_stream():
        while True:
            run = get_run(build)
            st = PROGRESS.get(build)
            if st:
                payload = {"phase": st["phase"], "pct": st["pct"], "message": st["message"], "done": st["done"]}
                yield f"data: {json.dumps(payload)}\n\n"
                if st.get("done"):
                    break
                await asyncio.sleep(0.35)
                continue

            if run is None:
                yield f"data: {json.dumps({'phase': 'error', 'pct': 0, 'message': 'Run not found', 'done': True})}\n\n"
                break
            if run.status == "complete":
                yield f"data: {json.dumps({'phase': 'complete', 'pct': 100, 'message': 'Complete', 'done': True})}\n\n"
                break
            if run.status == "failed":
                yield f"data: {json.dumps({'phase': 'failed', 'pct': 0, 'message': run.error or 'Failed', 'done': True})}\n\n"
                break
            yield f"data: {json.dumps({'phase': 'running', 'pct': 0, 'message': 'Starting…', 'done': False})}\n\n"
            await asyncio.sleep(0.35)

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "Connection": "keep-alive", "X-Accel-Buffering": "no"},
    )


@router.get("/{build}/charts")
async def list_charts(build: BuildId) -> list[str]:
    """List chart PNG filenames."""
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")
    return list_chart_filenames(build)


@router.get("/{build}/charts/{name}")
async def get_chart_png(build: BuildId, name: str) -> FileResponse:
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")
    try:
        path = chart_path(build, name)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if not path.is_file():
        raise HTTPException(status_code=404, detail="Chart not found")
    return FileResponse(path, media_type="image/png", filename=name)


@router.get("/{build}/players")
async def list_players(build: BuildId) -> list[dict[str, Any]]:
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")
    try:
        return player_summaries(build)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="outputs.csv not found yet") from exc


@router.get("/{build}/players/{pid}")
async def get_player(build: BuildId, pid: str) -> list[dict[str, Any]]:
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")
    try:
        rows = player_all_runs(build, pid)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="outputs.csv not found yet") from exc
    if not rows:
        raise HTTPException(status_code=404, detail="Player not found")
    return rows


@router.get("/{build}/godprogs")
async def list_godprogs(build: BuildId) -> list[dict[str, Any]]:
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")
    return godprogs_records(build)


@router.get("/{build}/download")
async def download_artifact(
    build: BuildId,
    artifact: Annotated[str, Query(description="analysis|csv")] = "analysis",
) -> FileResponse:
    if get_run(build) is None:
        raise HTTPException(status_code=404, detail="Run not found")
    if artifact in ("analysis", "xlsx", "analysis.xlsx"):
        path = analysis_xlsx_path(build)
        if not path.is_file():
            raise HTTPException(status_code=404, detail="analysis.xlsx not found")
        return FileResponse(
            path,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            filename="analysis.xlsx",
        )
    if artifact in ("csv", "raw", "outputs.csv"):
        path = raw_outputs_csv_path(build)
        if not path.is_file():
            raise HTTPException(status_code=404, detail="outputs.csv not found")
        return FileResponse(path, media_type="text/csv", filename="outputs.csv")
    raise HTTPException(status_code=422, detail="Invalid file query (use analysis or csv)")


@router.delete("/{build}")
async def delete_sim(build: BuildId) -> dict[str, bool]:
    """Delete a run output directory."""
    PROGRESS.pop(build, None)
    if not delete_run(build):
        raise HTTPException(status_code=404, detail="Run not found")
    return {"ok": True}


@router.get("/{build}")
async def get_sim(build: BuildId) -> RunMetadata:
    """Single run metadata."""
    run = get_run(build)
    if run is None:
        raise HTTPException(status_code=404, detail="Run not found")
    return run

File: api/src/routes/sims.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import { availableParallelism } from "node:os";
import type { FastifyInstance, FastifyReply } from "fastify";
import { z } from "zod";
import { isValidBuildId } from "../buildId.js";
import { outputsRoot } from "../paths.js";
import * as engineAdapter from "../services/engineAdapter.js";
import { PROGRESS, runSimulationJob } from "../services/runner.js";
import * as storage from "../services/storage.js";
import {
  generateTeaminfo,
  validateTeaminfo,
  InvalidTeaminfoError,
} from "../services/teaminfo.js";
import {
  analysisDashboardPath,
  analysisXlsxPath,
  ArtifactNotFoundError,
  chartPath,
  comparisonCacheDir,
  comparisonDashboardPath,
  godprogsRecords,
  InvalidChartNameError,
  listChartFilenames,
  playerAllRuns,
  playerSummaries,
  rawOutputsCsvPath,
} from "../services/simArtifacts.js";
import { runPythonComparison } from "../services/analysisPython.js";
import {
  MarkersNotFoundError,
  getAnalysisData,
  getComparisonData,
  getComparisonScorecard,
} from "../services/analysisExtract.js";
import type { SimProgressPayload } from "../types.js";

export interface SimsRouteOptions {
  scheduleBackground: (task: () => void | Promise<void>) => void;
}

function delay(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

function utcNowIso(): string {
  return new Date().toISOString().replace(/\+00:00$/, "Z");
}

function formatBuildId(d: Date): string {
  const p = (n: number, w = 2) => String(n).padStart(w, "0");
  return (
    `${d.getUTCFullYear()}${p(d.getUTCMonth() + 1)}${p(d.getUTCDate())}` +
    `${p(d.getUTCHours())}${p(d.getUTCMinutes())}${p(d.getUTCSeconds())}`
  );
}

function newBuildId(): string {
  return formatBuildId(new Date());
}

/** Parse a second-precision CalVer build id back into a UTC Date. */
function buildIdToDate(build: string): Date {
  return new Date(
    Date.UTC(
      Number(build.slice(0, 4)),
      Number(build.slice(4, 6)) - 1,
      Number(build.slice(6, 8)),
      Number(build.slice(8, 10)),
      Number(build.slice(10, 12)),
      Number(build.slice(12, 14)),
    ),
  );
}

/** Nudge a build id forward by one second, keeping it a valid CalVer id. */
function bumpBuildId(build: string): string {
  const d = buildIdToDate(build);
  d.setUTCSeconds(d.getUTCSeconds() + 1);
  return formatBuildId(d);
}

/**
 * Reserve a unique run directory with an exclusive mkdir (no recursive create).
 * Retries by bumping the CalVer id when the candidate already exists — prevents
 * two concurrent POSTs from sharing the same second-precision build dir.
 */
async function allocateBuildDir(startFrom?: string): Promise<{ build: string; out: string }> {
  await fsp.mkdir(outputsRoot(), { recursive: true });
  let candidate = startFrom ?? newBuildId();
  for (let attempt = 0; attempt < 32; attempt++) {
    const out = path.join(outputsRoot(), candidate);
    try {
      await fsp.mkdir(out);
      return { build: candidate, out };
    } catch (e) {
      const code = (e as NodeJS.ErrnoException).code;
      if (code !== "EEXIST") throw e;
      candidate = bumpBuildId(candidate);
    }
  }
  throw new Error("Could not allocate a unique build id");
}

async function removeRunDirs(...dirs: readonly string[]): Promise<void> {
  await Promise.all(
    dirs.map((dir) => fsp.rm(dir, { recursive: true, force: true }).catch(() => {})),
  );
}

function defaultNWorkers(requested: number | null | undefined): number {
  if (requested != null) return requested;
  // Requires Node.js >= 18.14.0 for os.availableParallelism
  const cpu =
    typeof availableParallelism === "function" ? availableParallelism() : 4;
  return Math.max(cpu - 1, 1);
}

/** Progression-script versions the engine can run. Default is the adopted v4.3. */
export const PROGRESSION_VERSIONS = ["v41", "v43"] as const;
export type ProgressionVersion = (typeof PROGRESSION_VERSIONS)[number];

/** Human display label for a progression version id. */
function versionLabel(version: string): string {
  return version === "v43" ? "v4.3" : version === "v41" ? "v4.1" : version;
}

const SimCreateBodySchema = z.object({
  teams: z.array(z.string()).default([]),
  seed: z.number().int().default(69),
  runs: z.number().int().positive().default(500),
  n_workers: z.number().int().positive().nullable().optional(),
  version: z.enum(PROGRESSION_VERSIONS).default("v43"),
  compare: z.boolean().default(true),
});

/** The progression version that isn't the selected one (binary enum today). */
function otherVersion(version: ProgressionVersion): ProgressionVersion {
  return PROGRESSION_VERSIONS.find((v) => v !== version) ?? version;
}

type SimCreateBody = z.infer<typeof SimCreateBodySchema>;

function parseSimCreateBody(json: string): SimCreateBody {
  const data = JSON.parse(json);
  return SimCreateBodySchema.parse(data);
}

function badRequest(reply: FastifyReply, status: number, detail: string) {
  return reply.status(status).send({ detail });
}

async function rejectAndCleanup(
  reply: FastifyReply,
  dirs: readonly string[],
  status: number,
  detail: string,
) {
  await removeRunDirs(...dirs);
  return badRequest(reply, status, detail);
}

async function validateBuildIdHandler(
  request: { params: unknown },
  reply: FastifyReply,
) {
  const { build } = request.params as { build: string };
  if (!isValidBuildId(build)) {
    return badRequest(reply, 422, "Invalid build id");
  }
  if (storage.getRun(build) == null) {
    return reply.status(404).send({ detail: "Run not found" });
  }
}

export async function registerSimsRoutes(
  fastify: FastifyInstance,
  opts: SimsRouteOptions,
): Promise<void> {
  fastify.get("/api/sims", async (_request, reply) => {
    return reply.send(storage.listRuns());
  });

  fastify.post("/api/sims", async (request, reply) => {
    let exportPath: string | null = null;
    let teaminfoPath: string | null = null;
    let configStr = "";

    const { build, out } = await allocateBuildDir();
    const allocatedDirs: string[] = [out];

    const tempFiles: string[] = [];
    try {
      for await (const part of request.parts()) {
        if (part.type === "file") {
          if (part.fieldname === "export") {
            const tempPath = path.join(out, ".export.tmp");
            tempFiles.push(tempPath);
            await pipeline(part.file, fs.createWriteStream(tempPath));
            const finalPath = path.join(out, "export.json");
            await fsp.rename(tempPath, finalPath);
            exportPath = finalPath;
            tempFiles.splice(tempFiles.indexOf(tempPath), 1);
          } else if (part.fieldname === "teaminfo") {
            const tempPath = path.join(out, ".teaminfo.tmp");
            tempFiles.push(tempPath);
            await pipeline(part.file, fs.createWriteStream(tempPath));
            const finalPath = path.join(out, "teaminfo.json");
            await fsp.rename(tempPath, finalPath);
            teaminfoPath = finalPath;
            tempFiles.splice(tempFiles.indexOf(tempPath), 1);
          }
        } else if (part.fieldname === "config") {
          configStr = String(part.value ?? "");
        }
      }

      if (!exportPath) {
        return rejectAndCleanup(reply, allocatedDirs, 422, "export file is empty");
      }

      const exportStat = await fsp.stat(exportPath);
      if (exportStat.size === 0) {
        return rejectAndCleanup(reply, allocatedDirs, 422, "export file is empty");
      }

      let body: SimCreateBody;
      try {
        body = parseSimCreateBody(configStr);
      } catch (e) {
        return rejectAndCleanup(
          reply,
          allocatedDirs,
          422,
          `config is not valid JSON: ${String(e)}`,
        );
      }

      const n_workers = defaultNWorkers(body.n_workers);
      if (n_workers < 1) {
        return rejectAndCleanup(reply, allocatedDirs, 422, "n_workers must be >= 1");
      }

      let exportData: Record<string, unknown>;
      try {
        const exportBuf = await fsp.readFile(exportPath, "utf8");
        exportData = JSON.parse(exportBuf) as Record<string, unknown>;
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        return rejectAndCleanup(reply, allocatedDirs, 422, `export is not valid JSON: ${msg}`);
      }

      let teaminfoSource: "generated" | "user";
      let teaminfoMap: Record<string, string>;

      if (teaminfoPath != null) {
        const teaminfoStat = await fsp.stat(teaminfoPath);
        if (teaminfoStat.size === 0) {
          return rejectAndCleanup(reply, allocatedDirs, 422, "teaminfo file is empty");
        }
        try {
          const teaminfoBuf = await fsp.readFile(teaminfoPath, "utf8");
          const rawTeaminfo = JSON.parse(teaminfoBuf);
          teaminfoMap = validateTeaminfo(rawTeaminfo);
        } catch (e) {
          if (e instanceof InvalidTeaminfoError) {
            return rejectAndCleanup(reply, allocatedDirs, 400, e.message);
          }
          if (e instanceof SyntaxError) {
            return rejectAndCleanup(
              reply,
              allocatedDirs,
              400,
              `teaminfo.json is not valid JSON: ${e.message}`,
            );
          }
          throw e;
        }
        teaminfoSource = "user";
      } else {
        teaminfoMap = generateTeaminfo(exportData);
        teaminfoSource = "generated";
      }

      if (!teaminfoPath) {
        teaminfoPath = path.join(out, "teaminfo.json");
      }
      await fsp.writeFile(
        teaminfoPath,
        JSON.stringify(teaminfoMap, null, 2),
        "utf8",
      );

      const exportMeta = exportData.meta as Record<string, unknown> | undefined;
      const exportTitle = typeof exportMeta?.title === "string" ? exportMeta.title : null;

      interface PairFields {
        pair_id: string;
        pair_role: "primary" | "baseline";
        paired_with: string;
      }

      function buildMeta(
        runBuild: string,
        version: ProgressionVersion,
        pair: PairFields | null,
      ): Record<string, unknown> {
        return {
          build: runBuild,
          // Requested version recorded now; runner patches script_version + progression
          // from the engine's own metadata post-run (executed truth).
          requested_version: version,
          script_version: versionLabel(version),
          engine_build: engineAdapter.engineBuildVersion(),
          teams: body.teams,
          seed: body.seed,
          runs: body.runs,
          n_workers,
          export_file: `outputs/${runBuild}/export.json`,
          export_title: exportTitle,
          teaminfo_file: `outputs/${runBuild}/teaminfo.json`,
          teaminfo_source: teaminfoSource,
          status: "running",
          started_at: utcNowIso(),
          completed_at: null,
          player_count: null,
          analysis_engine: null,
          error: null,
          ...(pair ?? {}),
        };
      }

      function scheduleRun(
        runBuild: string,
        runExportPath: string,
        runTeaminfoPath: string,
        version: ProgressionVersion,
      ) {
        opts.scheduleBackground(() =>
          runSimulationJob(
            runBuild,
            runExportPath,
            runTeaminfoPath,
            body.teams,
            body.seed,
            body.runs,
            n_workers,
            version,
          ),
        );
      }

      if (body.compare) {
        // Auto-comparison: a second run with the OTHER version, same inputs, its own
        // run dir. The selected version is primary; the other is the baseline.
        const baselineVersion = otherVersion(body.version);
        // Start from the next second so we never reuse the primary id; exclusive mkdir
        // then retries further if that slot is already taken by another request.
        const { build: baselineBuild, out: baselineOut } = await allocateBuildDir(
          bumpBuildId(build),
        );
        allocatedDirs.push(baselineOut);
        const pairId = build;

        const baselineExportPath = path.join(baselineOut, "export.json");
        const baselineTeaminfoPath = path.join(baselineOut, "teaminfo.json");
        await fsp.copyFile(exportPath, baselineExportPath);
        await fsp.copyFile(teaminfoPath, baselineTeaminfoPath);

        const primaryMeta = buildMeta(build, body.version, {
          pair_id: pairId,
          pair_role: "primary",
          paired_with: baselineBuild,
        });
        const baselineMeta = buildMeta(baselineBuild, baselineVersion, {
          pair_id: pairId,
          pair_role: "baseline",
          paired_with: build,
        });
        try {
          await fsp.writeFile(
            path.join(out, "metadata.json"),
            JSON.stringify(primaryMeta, null, 2),
            "utf8",
          );
          await fsp.writeFile(
            path.join(baselineOut, "metadata.json"),
            JSON.stringify(baselineMeta, null, 2),
            "utf8",
          );
        } catch (e) {
          const msg = e instanceof Error ? e.message : String(e);
          return rejectAndCleanup(
            reply,
            allocatedDirs,
            500,
            `Failed to write run metadata: ${msg}`,
          );
        }

        // Primary first so callers/tests see the selected version as the first job.
        scheduleRun(build, exportPath, teaminfoPath, body.version);
        scheduleRun(baselineBuild, baselineExportPath, baselineTeaminfoPath, baselineVersion);

        return reply.send({ build, compare_build: baselineBuild, pair_id: pairId });
      }

      const meta = buildMeta(build, body.version, null);
      try {
        await fsp.writeFile(path.join(out, "metadata.json"), JSON.stringify(meta, null, 2), "utf8");
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        return rejectAndCleanup(
          reply,
          allocatedDirs,
          500,
          `Failed to write run metadata: ${msg}`,
        );
      }

      scheduleRun(build, exportPath, teaminfoPath, body.version);

      return reply.send({ build });
    } catch (err) {
      await removeRunDirs(...allocatedDirs);
      // Clean up temp files on error
      for (const tmpFile of tempFiles) {
        await fsp.rm(tmpFile, { force: true }).catch(() => {});
      }
      throw err;
    } finally {
      // Clean up any remaining temp files
      for (const tmpFile of tempFiles) {
        await fsp.rm(tmpFile, { force: true }).catch(() => {});
      }
    }
  });

  fastify.get("/api/sims/:build/progress", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params as { build: string };

    async function* eventStream(): AsyncGenerator<string> {
      while (true) {
        const run = storage.getRun(build);
        const st = PROGRESS.get(build);
        if (st) {
          const payload: SimProgressPayload = {
            phase: st.phase,
            pct: st.pct,
            message: st.message,
            done: st.done,
          };
          yield `data: ${JSON.stringify(payload)}\n\n`;
          if (st.done) break;
          await delay(350);
          continue;
        }
        if (run == null) {
          yield `data: ${JSON.stringify({
            phase: "error",
            pct: 0,
            message: "Run not found",
            done: true,
          })}\n\n`;
          break;
        }
        if (run.status === "complete") {
          yield `data: ${JSON.stringify({
            phase: "complete",
            pct: 100,
            message: "Complete",
            done: true,
          })}\n\n`;
          break;
        }
        if (run.status === "failed") {
          yield `data: ${JSON.stringify({
            phase: "failed",
            pct: 0,
            message: run.error ?? "Failed",
            done: true,
          })}\n\n`;
          break;
        }
        yield `data: ${JSON.stringify({
          phase: "running",
          pct: 0,
          message: "Starting…",
          done: false,
        })}\n\n`;
        await delay(350);
      }
    }

    return reply
      .header("Content-Type", "text/event-stream")
      .header("Cache-Control", "no-cache")
      .header("Connection", "keep-alive")
      .header("X-Accel-Buffering", "no")
      .send(Readable.from(eventStream()));
  });

  fastify.get("/api/sims/:build/charts", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params as { build: string };
    return reply.send(listChartFilenames(build));
  });

  fastify.get("/api/sims/:build/charts/:name", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build, name } = request.params as { build: string; name: string };
    try {
      const p = chartPath(build, name);
      return reply.type("image/png").send(fs.createReadStream(p));
    } catch (e) {
      if (e instanceof InvalidChartNameError) {
        return badRequest(reply, 400, e.message);
      }
      if (e instanceof ArtifactNotFoundError) {
        return reply.status(404).send({ detail: "Chart not found" });
      }
      throw e;
    }
  });

  fastify.get("/api/sims/:build/players", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params as { build: string };
    try {
      return reply.send(playerSummaries(build));
    } catch {
      return reply.status(404).send({ detail: "outputs.csv not found yet" });
    }
  });

  fastify.get("/api/sims/:build/players/:pid", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build, pid } = request.params as { build: string; pid: string };
    try {
      const rows = playerAllRuns(build, pid);
      if (!rows.length) {
        return reply.status(404).send({ detail: "Player not found" });
      }
      return reply.send(rows);
    } catch {
      return reply.status(404).send({ detail: "outputs.csv not found yet" });
    }
  });

  fastify.get("/api/sims/:build/godprogs", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params as { build: string };
    return reply.send(godprogsRecords(build));
  });

  fastify.get<{
    Params: { build: string };
    Querystring: { artifact?: string };
  }>("/api/sims/:build/download", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params;
    const artifact = request.query.artifact ?? "analysis";
    if (["analysis", "xlsx", "analysis.xlsx"].includes(artifact)) {
      const p = analysisXlsxPath(build);
      if (!fs.existsSync(p)) {
        return reply.status(404).send({ detail: "analysis.xlsx not found" });
      }
      return reply
        .header(
          "Content-Type",
          "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        )
        .header("Content-Disposition", 'attachment; filename="analysis.xlsx"')
        .send(fs.createReadStream(p));
    }
    if (["csv", "raw", "outputs.csv"].includes(artifact)) {
      const p = rawOutputsCsvPath(build);
      if (!fs.existsSync(p)) {
        return reply.status(404).send({ detail: "outputs.csv not found" });
      }
      return reply
        .type("text/csv")
        .header("Content-Disposition", 'attachment; filename="outputs.csv"')
        .send(fs.createReadStream(p));
    }
    return badRequest(reply, 422, "Invalid file query (use analysis or csv)");
  });

  fastify.delete("/api/sims/:build", async (request, reply) => {
    const { build } = request.params as { build: string };
    if (!isValidBuildId(build)) return badRequest(reply, 422, "Invalid build id");
    PROGRESS.delete(build);
    if (!storage.deleteRun(build)) {
      return reply.status(404).send({ detail: "Run not found" });
    }
    return reply.send({ ok: true });
  });

  fastify.get("/api/sims/:build/analysis", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params as { build: string };
    const htmlPath = analysisDashboardPath(build);
    if (!fs.existsSync(htmlPath)) {
      return reply
        .status(404)
        .send({ detail: "analysis_dashboard.html not found for this run" });
    }
    return reply.type("text/html").send(fs.createReadStream(htmlPath));
  });

  /**
   * Shared validate + ensure-generated logic for the comparison routes.
   * Sends the error response and returns null on failure; returns the
   * comparison cache key once comparison_dashboard.html exists.
   */
  async function ensureComparison(
    buildsQuery: string | undefined,
    reply: FastifyReply,
  ): Promise<string | null> {
    const rawBuilds = (buildsQuery ?? "")
      .split(",")
      .map((b) => b.trim())
      .filter((b) => b.length > 0);
    const builds = [...new Set(rawBuilds)];

    if (builds.length < 2) {
      badRequest(reply, 400, "Provide at least 2 build ids to compare");
      return null;
    }
    for (const b of builds) {
      if (!isValidBuildId(b)) {
        badRequest(reply, 422, `Invalid build id: ${b}`);
        return null;
      }
    }
    for (const b of builds) {
      const run = storage.getRun(b);
      if (run == null) {
        reply.status(404).send({ detail: `Run not found: ${b}` });
        return null;
      }
      if (run.status !== "complete") {
        reply.status(409).send({ detail: `Run not complete: ${b}` });
        return null;
      }
    }

    // Cache keyed by the sorted build-id set: order-independent, reused on repeat.
    const key = [...builds].sort().join("_");
    const htmlPath = comparisonDashboardPath(key);
    if (!fs.existsSync(htmlPath)) {
      const runDirs = builds.map((b) => path.join(outputsRoot(), b));
      try {
        await runPythonComparison(runDirs, comparisonCacheDir(key));
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        reply.status(500).send({ detail: `Comparison generation failed: ${msg}` });
        return null;
      }
      if (!fs.existsSync(htmlPath)) {
        reply.status(500).send({ detail: "Comparison dashboard was not produced" });
        return null;
      }
    }
    return key;
  }

  fastify.get<{ Querystring: { builds?: string } }>(
    "/api/sims/compare",
    async (request, reply) => {
      const key = await ensureComparison(request.query.builds, reply);
      if (key == null) return reply;
      return reply
        .type("text/html")
        .send(fs.createReadStream(comparisonDashboardPath(key)));
    },
  );

  fastify.get<{ Querystring: { builds?: string } }>(
    "/api/sims/compare-data",
    async (request, reply) => {
      const key = await ensureComparison(request.query.builds, reply);
      if (key == null) return reply;
      try {
        const data = getComparisonData(key);
        return reply.send({
          ...data,
          engine: "python",
          builds: key.split("_"),
          scorecard: getComparisonScorecard(key),
        });
      } catch (e) {
        if (e instanceof MarkersNotFoundError) {
          return reply.status(409).send({
            detail: "Native comparison unavailable: dashboard markers not found",
          });
        }
        throw e;
      }
    },
  );

  fastify.get(
    "/api/sims/:build/analysis-data",
    { preHandler: validateBuildIdHandler },
    async (request, reply) => {
      const { build } = request.params as { build: string };
      const htmlPath = analysisDashboardPath(build);
      if (!fs.existsSync(htmlPath)) {
        return reply
          .status(404)
          .send({ detail: "analysis_dashboard.html not found for this run" });
      }
      const run = storage.getRun(build);
      if (run?.analysis_engine === "fallback") {
        return reply.status(409).send({
          detail: "Native analysis unavailable: run used the fallback analysis engine",
        });
      }
      try {
        const data = getAnalysisData(build);
        return reply.send({ ...data, engine: "python", build });
      } catch (e) {
        if (e instanceof MarkersNotFoundError) {
          return reply.status(409).send({
            detail: "Native analysis unavailable: dashboard markers not found",
          });
        }
        throw e;
      }
    },
  );

  fastify.get("/api/sims/:build", { preHandler: validateBuildIdHandler }, async (request, reply) => {
    const { build } = request.params as { build: string };
    const run = storage.getRun(build);
    return reply.send(run);
  });
}
